ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+        else:
+
+            async def _func(ep):  # coroutine wrapper
+                func(ep)
+
+            await _func(ep)
+
+    asyncio.ensure_future(run(ucp_endpoint, ctx, worker, func, guarantee_msg_order))
+
+
+async def _non_blocking_mode(weakref_ctx):
+    """This help function maintains a UCX progress loop.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        del ctx
+        await asyncio.sleep(0)
+
+
+async def _arm_worker(weakref_ctx, rsock, event_loop):
+    """This help function args the worker.
+    Notice, it only keeps a weak reference to `ApplicationContext`, which makes it
+    possible to call `ucp.reset()` even when this loop is running.
+    """
+
+    # When arming the worker, the following must be true:
+    #  - No more progress in UCX (see doc of ucp_worker_arm())
+    #  - All asyncio tasks that isn't waiting on UCX must be executed
+    #    so that the asyncio's next state is epoll wait.
+    #    See <https://github.com/rapidsai/ucx-py/issues/413>
+    while True:
+        ctx = weakref_ctx()
+        if ctx is None:
+            return
+        ctx.progress()
+        # This IO task returns when all non-IO tasks are finished.
+        await event_loop.sock_recv(rsock, 1)
+        if ctx._worker.arm():
+            # At this point we know that asyncio's next state is
+            # epoll wait.
+            break
+        del ctx
+
+
+class ApplicationContext:
+    def __init__(self, config_dict={}, blocking_progress_mode=None):
+        self.event_loops_binded_for_progress = set()
+        self.progress_tasks = []
+        self.dangling_arm_task = None
+        self.epoll_fd = -1
+        self._ctx = None
+        self._worker = None
+        if blocking_progress_mode is not None:
+            self.blocking_progress_mode = blocking_progress_mode
+        elif "UCXPY_NON_BLOCKING_MODE" in environ:
+            self.blocking_progress_mode = False
+        else:
+            self.blocking_progress_mode = True
+
+        # Create context and worker, whic might fail
+        self._ctx = ucx_api.UCXContext(config_dict)
+        self._worker = ucx_api.UCXWorker(self._ctx)
+
+        if self.blocking_progress_mode:
+            self.epoll_fd = self._worker.init_blocking_progress_mode()
+
+    def __del__(self):
+        for task in self.progress_tasks:
+            task.cancel()
+        if self.dangling_arm_task is not None:
+            self.dangling_arm_task.cancel()
+        # Notice, worker and context might never have been created
+        if self._worker is not None:
+            self._worker.destroy()
+        if self._ctx is not None:
+            self._ctx.destroy()
+        if self.blocking_progress_mode and self.epoll_fd != -1:
+            close_fd(self.epoll_fd)
```

Comment:
Is it possible to move the cleanup stuff into Cython (for example) to avoid the need for an explicit `__del__`?

Reply:
I thought it was better to have it in `__del__()` than `__dealloc__()` ?

Reply:
Well they both have their issues. `__dealloc__` relies on the Python interpreter to run the cleanup step, which may not happen during a bad shutdown. Though the same could be said of `__del__`. However `__del__` has [even more issues]( https://stackoverflow.com/a/1481512/3877089 ). So unless we are particularly worried about the bad shutdown case, would go the `__dealloc__` route.